### PR TITLE
Feature/openstack

### PR DESCRIPTION
**Pull request by mateimicu**

### CODE REVIEW

File: arestor/api/__init__.py [hunk 1]
```diff
@@ -45,6 +45,6 @@ def config(cls):
                 'server.thread_pool': CONFIG.api.thread_pool,
```
Comment (ader1990) [inline arestor/api/__init__.py:51]: being an API change, this should have a commit message that presents the changes

**Review by ader1990 — CHANGES_REQUESTED** [review]

File: arestor/common/tools.py [hunk 2]
```diff
@@ -99,10 +100,36 @@ def _process_content(secret):
...
+        except ValueError:
+            LOG.error("Failed to check the signature.")
+            return False
+
+        if not isinstance(content, dict) or "timestamp" not in content:
+            LOG.error("Invalid signature provided.")
+            return False
+
+        if content["timestamp"] != timestamp:
+            LOG.error("Malformated signature provided.")
```
Comment (ader1990) [inline arestor/common/tools.py]: Malformed*

**Review by ader1990 — CHANGES_REQUESTED** [review]
plus add a more informational commit message

**Review by ader1990 — CHANGES_REQUESTED** [review]
add a better commit message

**Review by ader1990 — CHANGES_REQUESTED** [review]
Add a better commit message

File: arestor/api/admin/resource.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+            # Prepare the representation of the received resource.
+            resource = connection.hgetall(resource_id)
+            if not resource:
+                response["meta"]["status"] = False
+                response["meta"]["verbose"] = "Resource not found"
+
+            response["content"] = resource
+            return response
+
+        if not any([resource_id, namespace, client_id, resource]):
```
Comment (paulaCrismaru) [inline arestor/api/admin/resource.py]: unnecessary. same result as L61-L68

File: arestor/api/v1/openstack.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+
+class _OpenStackResource(base_api.Resource):
+    """Base class for OpenStack resources."""
+
+    exposed = True
+
+    def _get_openstack_data(self, name, field=None):
+        """Retrieve the required resource from the Openstack namespace."""
+        data = None
```
Comment (paulaCrismaru) [inline arestor/api/v1/openstack.py:36]: Should it be None or empty string?

File: arestor/api/v1/openstack.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+
+            if field == "data":
+                data = json.loads(data)
+        except (exception.NotFound, ValueError):
+            pass
+        return data
+
+    def _set_openstack_data(self, name, field=None, value=None):
+        """Set the required resource from the Openstack namespace."""
+        data = None
```
Comment (paulaCrismaru) [inline arestor/api/v1/openstack.py:49]: same

File: arestor/api/v1/openstack.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+    exposed = True
+    resources = [
+        ("user_data", _UserdataResource),
+        ("meta_data_json", _MetadataResource),
+    ]
+
+
+class _LegacyVersion(base_api.BaseAPI):
+
+    """Containe for all the resources from the legacy version of the API."""
```
Comment (paulaCrismaru) [inline arestor/api/v1/openstack.py]: Container*

File: arestor/client/arestor_client.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+        data = self.resource(key).get("data", {})
+        try:
+            data = json.loads(data)
+        except ValueError:
+            pass
+        return data
+
+    def get_url(self):
+        """Return the url for this client."""
+        url = self._base_url + "/v1/" + self._client_id + "/" + self._namespace
```
Comment (paulaCrismaru) [inline arestor/client/arestor_client.py]: you should use  url_parse.urljoin

**Review by paulaCrismaru — CHANGES_REQUESTED** [review]

File: arestor/api/admin/__init__.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
+# Copyright 2016 Cloudbase Solutions Srl
```
Comment (ader1990) [inline arestor/api/admin/__init__.py]: 2017*

File: arestor/api/admin/resource.py [hunk 9]
```diff
@@ -0,0 +1,123 @@
+# Copyright 2016 Cloudbase Solutions Srl
```
Comment (ader1990) [inline arestor/api/admin/resource.py]: ditto

File: arestor/api/__init__.py [hunk 10]
```diff
@@ -19,17 +19,23 @@
 import cherrypy
```
Comment (ader1990) [inline arestor/api/__init__.py:19]: add a commit description

**Review by ader1990 — CHANGES_REQUESTED** [review]

File: arestor/client/__init__.py [hunk 11]
```diff
@@ -0,0 +1,13 @@
+# Copyright 2016 Cloudbase Solutions Srl
```
Comment (ader1990) [inline arestor/client/__init__.py]: ditto

File: arestor/client/base.py [hunk 12]
```diff
@@ -0,0 +1,79 @@
+# Copyright 2016 Cloudbase Solutions Srl
```
Comment (ader1990) [inline arestor/client/base.py]: ditto

File: arestor/client/base.py [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+        try:
+            response = requests.request(method=method, url=url,
+                                        data=data)
+        except requests.RequestException as exc:
+            raise exception.ClientError(exc)
+
+        return response
+
+    def get(self, resource):
+        """View the required resource."""
```
Comment (ader1990) [inline arestor/client/base.py]: Get the required*

File: arestor/client/base.py [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+class Client(BaseClient):
+
+    """Basic Arestor API client."""
+
+    def __init__(self, base_url, api_key, secret):
+        super(Client, self).__init__(base_url)
+        self._key = api_key
+        self._cipher = arestor_util.AESCipher(secret)
+
+    def _get_parameters(self):
```
Comment (ader1990) [inline arestor/client/base.py]: _get_auth_params -- name the methods according to what they do

File: arestor/common/exception.py [hunk 15]
```diff
@@ -49,6 +49,14 @@ class CliError(ArestorException):
     template = "Something went wrong during the procesing of command line."
 
 
+class ClientError(ArestorException):
+
+    """Something went wrong during communication with the API."""
```
Comment (ader1990) [inline arestor/common/exception.py]: This should describe the class, and not be an error message.

File: arestor/common/exception.py [hunk 16]
```diff
@@ -49,6 +49,14 @@ class CliError(ArestorException):
...
+
+    template = ("Something went wrong during "
```
Comment (ader1990) [inline arestor/common/exception.py]: Failed to communicate with the Arestor API.

**Review by ader1990 — CHANGES_REQUESTED** [review]

File: arestor/client/resource.py [hunk 17]
```diff
@@ -0,0 +1,127 @@
+# Copyright 2016 Cloudbase Solutions Srl
```
Comment (ader1990) [inline arestor/client/resource.py]: ditto

File: arestor/client/resource.py [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+        except ValueError:
+            raise exception.ClientError(msg="Malformed response.")
+
+        if not data["meta"]["status"]:
+            raise exception.ClientError(msg=data["meta"]["verbose"])
+
+        return data["content"]
+
+    def update_resource(self, resource_id, content):
+        """Update the content of the required resource."""
```
Comment (ader1990) [inline arestor/client/resource.py]: of the given*

File: arestor/client/resource.py [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+        except ValueError:
+            raise exception.ClientError(msg="Malformed response.")
+
+        if not resource["meta"]["status"]:
+            raise exception.ClientError(msg=resource["meta"]["verbose"])
+
+        return resource["content"]
+
+    def delete_resource(self, resource_id):
+        """Delete the required resource."""
```
Comment (ader1990) [inline arestor/client/resource.py]: the given*

File: arestor/client/resource.py (hunk 17 above)
Comment (ader1990) [inline arestor/client/resource.py]: add a description for the commit

**Review by ader1990 — CHANGES_REQUESTED** [review]

File: arestor/client/arestor_client.py [hunk 20]
```diff
@@ -0,0 +1,125 @@
+# Copyright 2016 Cloudbase Solutions Srl
```
Comment (ader1990) [inline arestor/client/arestor_client.py]: ditto1

Comment (ader1990) [reply arestor/client/arestor_client.py]: ditto2 (add a better description to what you commited)

File: arestor/client/arestor_client.py [hunk 21]
```diff
@@ -0,0 +1,125 @@
...
+
+    def get_password(self):
+        return self._get_resource("password")
+
+    def get_ssh_pubkeys(self):
+        ssh_keys = self._get_resource("public_keys")
+        return ssh_keys
+
+    def delete_all_data(self):
+        """Delete all meta_data for this client_id."""
```
Comment (ader1990) [inline arestor/client/arestor_client.py]: for the current client_id

**Review by ader1990 — CHANGES_REQUESTED** [review]

File: arestor/api/v1/openstack.py [hunk 22]
```diff
@@ -124,6 +124,10 @@ class _LegacyVersion(base_api.BaseAPI):
     exposed = True
     resources = [
         ("password", _PasswordResource),
+        # NOTE(mmicu): Cloudbase-Init will check if this endpoint is
+        # available by requesting the `/meta_data.json` file,
+        # this could be emply
```
Comment (ader1990) [inline arestor/api/v1/openstack.py]: empty*

**Review by ader1990 — CHANGES_REQUESTED** [review]

File: arestor/api/v1/openstack.py [hunk 23]
```diff
@@ -13,6 +13,7 @@
 #    under the License.
```
Comment (ader1990) [inline arestor/api/v1/openstack.py]: ditto2

**Review by ader1990 — CHANGES_REQUESTED** [review]

**Review by ader1990 — APPROVED** [review]

File: arestor/api/__init__.py (hunk 10 above)
Comment (mateimicu) [reply arestor/api/__init__.py:19]: Done

File: arestor/api/v1/openstack.py (hunk 4 above)
Comment (mateimicu) [reply arestor/api/v1/openstack.py:36]: This is a good question. The API works either way but the consumer(Cloudbase-Init in this case) will do checks like this:
```
if not random_metadata:
     # ...
``` 
A empty string and None will yield the same result.

File: arestor/api/__init__.py (hunk 10 above)
Comment (ader1990) [reply arestor/api/__init__.py:19]: can't see it

File: arestor/api/__init__.py (hunk 1 above)
Comment (mateimicu) [reply arestor/api/__init__.py:51]: Done

**Review by ader1990 — APPROVED** [review]

**Pull request merged**